generator/src/main/java/com/scottlogic/deg/generator/generation/RegexStringGenerator.java

```diff
@@ -424,35 +422,89 @@ void setMaxChar(char maxChar) {
     private class FiniteStringAutomatonIterator implements Iterator<String> {
 
         private final RegexStringGenerator stringGenerator;
+        private final long matches;
         private int currentIndex;
+        private String currentValue;
 
         FiniteStringAutomatonIterator(RegexStringGenerator stringGenerator) {
             this.stringGenerator = stringGenerator;
+            this.matches = stringGenerator.getValueCount();
             currentIndex = 0;
         }
 
+        /**
+         * <p>
+         * This function has been updated to only allow valid single 16-bit
+         * word UTF-8 characters to be output.
+         * </p>
+         * <p>
+         *     FIXME - This check will be removed if/when the dk.brics.automaton
+         *     library is fixed to support surrogate pairs,
+         * </p>
+         * <p>
+         *     issue #15 (https://github.com/cs-au-dk/dk.brics.automaton/issues/15)
+         *     has been raised on the dk.brics.automaton library
+         * </p>
+         * <p>
+         *     issue #537 has been created to track when the dk.brics.automaton library
+         *     is updated.
+         * </p>
+         */
         @Override
         public boolean hasNext() {
-            long matches = stringGenerator.getValueCount();
-            return currentIndex < matches;
+            if (currentValue != null) {
+                return true;
+            }
+            do {
+                currentIndex++; // starts at 1
+                if (currentIndex > matches) {
+                    return false;
+                }
+                currentValue = stringGenerator.getMatchedString(currentIndex);
+            } while (!containsValidUtf8Chars(currentValue));
```

Comment:
I know people might not like do{} while() loops, but it saves duplication of code

Reply:
I think do..while loops are fine, as long as there is a reason to their use